numpy/core/src/multiarray/getset.c

```diff
@@ -573,12 +583,12 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
         Py_DECREF(temp);
     }
 
+    Py_DECREF(PyArray_DESCR(self));
```

Comment:
On the other hand, leaving it where it was brings into question whether something was looking at a decref'd dtype.